### ISSUE TITLE
Create a bot to check if there is a CHANGES entry for new PRs

### ISSUE THREAD

**Issue by stefanvodita** [apache/lucene#13898]
### Description

We always have to remind ourselves and others to add a CHANGES entry. We can probably write a GitHub action to check for changes to that file on every new PR and leave a comment warning the author if that's missing.

**Comment by prudhvigodithi** [comment]
Here is the sample PR with change-log verifier action added https://github.com/prudhvigodithi/lucene/pull/1, we can add the same workflow to lucene repo. But I see the error `dangoslen/changelog-enforcer@v3 is not allowed to be used in apache/lucene. `.

**Comment by javanna** [comment]
Yes to this! It would be great to combine this with setting the milestone :)

**Comment by stefanvodita** [comment]
@prudhvigodithi - I opened #13898 inspired by your PR, tested it on my fork, and it's working correctly.
@javanna  - I like that idea! Maybe we can do that as a follow-up.

**Comment by stefanvodita** [comment]
I was looking at the workflow that @pseudo-nymous wrote. The [logs](https://github.com/apache/lucene/actions/runs/13597403773/job/38017259510) for #14276 say there is no CHANGES entry. We'll need to fix that. I've also added the PR number to the output (#14324), since I couldn't see it anywhere.

**Comment by pseudo-nymous** [comment]
Thanks for updating the script with PR number. I'll root cause and fix the issue.

**Comment by stefanvodita** [comment]
@pseudo-nymous, I've attempted a fix in #14355.

**Comment by dweiss** [comment]
> dangoslen/changelog-enforcer@v3 is not allowed to be used in apache/lucene

Apache projects are not allowed to use arbitrary actions for security reasons, please take a look at:
https://infra.apache.org/github-actions-policy.html#External%5C%20actions

**Comment by pseudo-nymous** [comment]
@stefanvodita  I have added a fix for it. Please take a look. https://github.com/apache/lucene/pull/14369

**Comment by stefanvodita** [comment]
I created a new GitHub account with its own Lucene fork and opened a [PR](https://github.com/stefanvodita/lucene/pull/9) against my old account to replicate the behaviour we were seeing. This worked well and I think I've been able to fix a few of the issues in #14606. We'll monitor after the fix is merged. For now, the workflow is still only logging and not actually performing write operations on PRs.

**Comment by stefanvodita** [comment]
I pushed a few more fixes in #14628. We have our first successful instance of the changelog being validated and a milestone extracted from it ([logs](https://github.com/apache/lucene/actions/runs/14906069230/job/41868625305))!

**Comment by jainankitk** [comment]
> > dangoslen/changelog-enforcer@v3 is not allowed to be used in apache/lucene
> 
> Apache projects are not allowed to use arbitrary actions for security reasons, please take a look at: https://infra.apache.org/github-actions-policy.html#External%5C%20actions

Okay, I was just curious about the resource usage policy and noticed some inconsistent information (calendar week should have more allowed runners):

* The average number of minutes a project uses per calendar week MUST NOT exceed the equivalent of 25 full-time runners (250,000 minutes, or 4,200 hours).
* The average number of minutes a project uses in any consecutive five-day period MUST NOT exceed the equivalent of 30 full-time runners (216,000 minutes, or 3,600 hours).

Wondering if we can fix this documentation inconsistency

**Comment by stefanvodita** [comment]
I've been monitoring the jobs after the most recent batch of fixes and I'm happy with the results. The only change that the bot got wrong was #14638 ([logs](https://github.com/apache/lucene/actions/runs/14941689255/job/41979703606)), which moves around a changelog entry from 11.0.0 to 10.3.0. The bot would still assign it milestone 11.0.0, but I think that's not a serious mistake and I'm willing to take a progress-not-perfection approach here. I opened #14644 to start assigning milestones automatically ([test](https://github.com/stefanvodita/lucene/actions/runs/14960617144/job/42022012969?pr=9)) and to post a comment on PRs that don't have a changelog entry ([test](https://github.com/stefanvodita/lucene/actions/runs/14960628233/job/42022036265?pr=9)). If there aren't objections, I will send an announcement to the dev list ahead of pushing this and enabling the bot.

**Comment by stefanvodita** [comment]
The bot has processed one PR (#14653) and I already see something I'm not sure I like - it will post a comment every time we push to the branch and are missing a CHANGES entry. I can see arguments on both sides, but I think I would prefer a single comment per PR. We could probably do that by adding a label to the PR after the first comment.

**Comment by dweiss** [comment]
One would be probably enough. I'm also not sure we should require all changes in changes.txt. The one in 14653 doesn't seem to be that important (it applies to build infrastructure, nobody looking at changes.txt will ever care about it?).

**Comment by rmuir** [comment]
Maybe we could look for a change to `*.java` as a heuristic?

This would prevent noise for stuff like build system updates, doc updates, dependabot updates, python dev scripts, etc?

**Comment by stefanvodita** [comment]
I thought about it some more. The way we envisioned PR authors choosing not to run the verifier is by adding the [`skip-changelog-check` label](https://github.com/apache/lucene/labels/skip-changelog-check). I should have mentioned this in the email I sent to the dev list.
Here is what I'd like to do next if it seems good to you as well.
1. Update the bot's comment to tell people they can use the label to skip the check on their PR. If they don't add the label, the bot will continue reminding us as the PR gets updated. If they do add the label, the bot will stop.
2. Eventually make the bot smarter like you are suggesting, so that it doesn't run on PRs that we know don't need a CHANGES entry. Maybe we would do this by unifying with the [module lebeler](https://github.com/apache/lucene/blob/main/.github/labeler.yml), which already looks at which files were touched.

**Comment by stefanvodita** [comment]
#14656 is what I was suggesting as an immediate step. I wonder if all contributors have permissions to add labels to their PRs though.

**Comment by msfroh** [comment]
What I've seen on OpenSearch (where we have a similar changelog verifier, with a skip-changelog label) is that you do need some level of repo permissions to be able to add labels. As far as I know, for Lucene that means being a committer.

That said, it's not a terrible to leave the "Is this worthy of a changelog entry?" decision up to the person who's going to merge the change, since it is a judgment call. I like Robert's suggestion of skipping the check by default if there are no `*.java` files, though. If we could default to adding the skip label so a human could remove it if so inclined, that would (IMO) strike the best balance between managing noise and flexibility. Of course, I imagine it would be really easy to end up in a situation where the action keeps adding the label back (because there are no `.java` files) and a human keeps removing it. 😄

**Comment by shubhamsrkdev** [comment]
> [#14656](https://github.com/apache/lucene/pull/14656) is what I was suggesting as an immediate step. I wonder if all contributors have permissions to add labels to their PRs though.

+1  - I couldn't/can't add this label or any labels on my PR.

**Comment by pseudo-nymous** [comment]
Github does not allow adding labels by non owners of a repo. Ref: [link](https://github.com/isaacs/github/issues/875)
